# 🛒 AdventureWorks ETL Project

# Customer Dimension ETL Pipeline

---

## 📌 Objective

The objective of this notebook is to perform the complete ETL (Extract, Transform, Load) process for the **Customers** dataset.

This notebook covers:

- Understanding the dataset
- Creating the Data Dictionary
- Data Profiling
- Data Quality Assessment
- Data Cleaning & Transformation
- Data Validation
- Exporting the cleaned dataset

The cleaned dataset will later be loaded into **MySQL** and used for building an interactive **Power BI Dashboard**.

---

# ETL Workflow

```text
Raw CSV
    │
    ▼
Extract
    │
    ▼
Data Profiling
    │
    ▼
Data Quality Assessment
    │
    ▼
Data Cleaning
    │
    ▼
Validation
    │
    ▼
Export Clean Dataset
```

---

# Dataset Information

| Attribute | Description |
|------------|------------|
| Dataset | AdventureWorks Customers |
| File Name | AdventureWorks_Customers.csv |
| Table Type | Dimension Table |
| Primary Key | CustomerKey |
| Records | 18,148 |
| Columns | 13 |

---

# Business Understanding

The Customers table stores demographic and personal information about customers.

It will be used for:

- Customer Segmentation
- Sales Analysis
- Revenue Analysis
- Customer Lifetime Value (CLV)
- Geographic Analysis
- Power BI Dashboards

This table will later be connected with the **Sales** table using **CustomerKey**.

---

# ETL Stages Covered

## 1. Extract

- Load raw CSV
- Verify successful import

---

## 2. Data Profiling

- Dataset shape
- Column names
- Data types
- Missing values
- Duplicate records
- Unique values
- Summary statistics
- Domain validation

---

## 3. Data Quality Assessment

Evaluate data based on:

- Completeness
- Uniqueness
- Validity
- Consistency
- Accuracy

---

## 4. Data Cleaning

Transformations include:

- Rename columns to snake_case
- Convert data types
- Handle missing values
- Standardize categorical values
- Standardize text formatting
- Remove unnecessary whitespace

---

## 5. Validation

Verify:

- Data types
- Missing values
- Duplicate records
- Domain values
- Primary Key uniqueness

---

## 6. Export

Export the cleaned dataset to:

```text
data/cleaned/customers_clean.csv
```

The raw dataset will remain unchanged inside:

```text
data/raw/
```

---

# Expected Output

At the end of this notebook, the Customers dataset will be:

- Clean
- Consistent
- Analysis-ready
- Ready for MySQL
- Ready for Power BI
- Ready for SQL analysis

---

# Technologies Used

- Python
- Pandas
- NumPy
- Jupyter Notebook
- MySQL (Next Stage)
- Power BI (Dashboard Stage)

---



# 1. Import Libraries and Load Dataset

In [239]:
import pandas as pd
import numpy as np

customers = pd.read_csv(
    "C:/Users/singh/OneDrive/Desktop/AdventureWorks_Analytics/data/raw/AdventureWorks_Customers.csv",
    encoding="latin1"
)

# 2. Data Profiling

In [240]:
customers.head()
customers.tail()

,CustomerKey,Prefix,FirstName,LastName,BirthDate,MaritalStatus,Gender,EmailAddress,AnnualIncome,TotalChildren,EducationLevel,Occupation,HomeOwner
18143,29479,MR.,TOMMY,TANG,7/4/1958,M,M,tommy2@adventure-works.com,"$30,000",1,Graduate Degree,Clerical,Y
18144,29480,MRS.,NINA,RAJI,11/10/1960,S,F,nina21@adventure-works.com,"$30,000",3,Graduate Degree,Clerical,Y
18145,29481,MR.,IVAN,SURI,1/5/1960,S,M,ivan0@adventure-works.com,"$30,000",3,Graduate Degree,Clerical,N
18146,29482,MR.,CLAYTON,ZHANG,3/5/1959,M,M,clayton0@adventure-works.com,"$30,000",3,Bachelors,Clerical,Y
18147,29483,MR.,JÉSUS,NAVARRO,12/8/1959,M,M,jésus9@adventure-works.com,"$30,000",0,Bachelors,Clerical,Y


In [241]:
customers.shape

(18148, 13)

In [242]:
customers.columns

Index(['CustomerKey', 'Prefix', 'FirstName', 'LastName', 'BirthDate',
       'MaritalStatus', 'Gender', 'EmailAddress', 'AnnualIncome',
       'TotalChildren', 'EducationLevel', 'Occupation', 'HomeOwner'],
      dtype='object')

In [243]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18148 entries, 0 to 18147
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   CustomerKey     18148 non-null  int64 
 1   Prefix          18018 non-null  object
 2   FirstName       18148 non-null  object
 3   LastName        18148 non-null  object
 4   BirthDate       18148 non-null  object
 5   MaritalStatus   18148 non-null  object
 6   Gender          18018 non-null  object
 7   EmailAddress    18148 non-null  object
 8   AnnualIncome    18148 non-null  object
 9   TotalChildren   18148 non-null  int64 
 10  EducationLevel  18148 non-null  object
 11  Occupation      18148 non-null  object
 12  HomeOwner       18148 non-null  object
dtypes: int64(2), object(11)
memory usage: 1.8+ MB


In [244]:
customers.isnull().sum()

CustomerKey         0
Prefix            130
FirstName           0
LastName            0
BirthDate           0
MaritalStatus       0
Gender            130
EmailAddress        0
AnnualIncome        0
TotalChildren       0
EducationLevel      0
Occupation          0
HomeOwner           0
dtype: int64

In [245]:
customers.duplicated().sum()

np.int64(0)

In [246]:
customers.describe()

,CustomerKey,TotalChildren
count,18148.000000,18148.000000
mean,20247.470465,1.843619
std,5336.245656,1.613203
min,11000.000000,0.000000
25%,15626.750000,0.000000
50%,20252.500000,2.000000
75%,24868.250000,3.000000
max,29483.000000,5.000000


In [247]:
customers.nunique()

CustomerKey       18148
Prefix                3
FirstName           666
LastName            372
BirthDate          8190
MaritalStatus         2
Gender                2
EmailAddress      18148
AnnualIncome         16
TotalChildren         6
EducationLevel        5
Occupation            5
HomeOwner             2
dtype: int64

In [248]:
customers["Gender"].value_counts(dropna=False)

Gender
M      9126
F      8892
NaN     130
Name: count, dtype: int64

In [249]:
customers["MaritalStatus"].value_counts(dropna=False)

MaritalStatus
M    9817
S    8331
Name: count, dtype: int64

In [250]:
customers["HomeOwner"].value_counts(dropna=False)

HomeOwner
Y    12260
N     5888
Name: count, dtype: int64

In [251]:
customers["EducationLevel"].value_counts(dropna=False)

EducationLevel
Bachelors              5261
Partial College        4966
High School            3241
Graduate Degree        3125
Partial High School    1555
Name: count, dtype: int64

In [252]:
customers["Occupation"].value_counts(dropna=False)

Occupation
Professional      5424
Skilled Manual    4501
Management        3011
Clerical          2859
Manual            2353
Name: count, dtype: int64

In [253]:
customers["AnnualIncome"].value_counts().sort_index()

AnnualIncome
$10,000      1135
$100,000      559
$110,000      459
$120,000      324
$130,000      501
$150,000      100
$160,000       91
$170,000      109
$20,000      1745
$30,000      2236
$40,000      2688
$50,000       661
$60,000      3072
$70,000      2313
$80,000      1325
$90,000       830
Name: count, dtype: int64

# 3. Data Quality Assessment

The Customers dataset was assessed before cleaning to identify data quality issues.

### Findings

- Missing values found in **Prefix (130)** and **Gender (130)**.
- No duplicate records found.
- No inconsistent categorical values detected.
- `BirthDate` requires conversion to **datetime**.
- `AnnualIncome` requires conversion to **numeric**.

### Conclusion

The dataset is in good condition with only minor data quality issues. The identified issues will be resolved during the data cleaning stage to prepare the dataset for SQL analysis and Power BI reporting.

# 4. Data Cleaning

In [254]:
customers.head()
customers.shape

(18148, 13)

In [255]:
customers.columns = (
    customers.columns
    .str.strip()
    .str.replace(" ","_")
    .str.replace(r"(?<!^)(?=[A-Z])", "_", regex=True)
    .str.lower()
    )

In [256]:
customers.columns


Index(['customer_key', 'prefix', 'first_name', 'last_name', 'birth_date',
       'marital_status', 'gender', 'email_address', 'annual_income',
       'total_children', 'education_level', 'occupation', 'home_owner'],
      dtype='object')

In [257]:
customers['birth_date'] = pd.to_datetime(
    customers['birth_date'],
    errors='coerce'
    )

In [258]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18148 entries, 0 to 18147
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   customer_key     18148 non-null  int64         
 1   prefix           18018 non-null  object        
 2   first_name       18148 non-null  object        
 3   last_name        18148 non-null  object        
 4   birth_date       18148 non-null  datetime64[ns]
 5   marital_status   18148 non-null  object        
 6   gender           18018 non-null  object        
 7   email_address    18148 non-null  object        
 8   annual_income    18148 non-null  object        
 9   total_children   18148 non-null  int64         
 10  education_level  18148 non-null  object        
 11  occupation       18148 non-null  object        
 12  home_owner       18148 non-null  object        
dtypes: datetime64[ns](1), int64(2), object(10)
memory usage: 1.8+ MB


In [259]:
customers.head()

,customer_key,prefix,first_name,last_name,birth_date,marital_status,gender,email_address,annual_income,total_children,education_level,occupation,home_owner
0,11000,MR.,JON,YANG,1966-04-08,M,M,jon24@adventure-works.com,"$90,000",2,Bachelors,Professional,Y
1,11001,MR.,EUGENE,HUANG,1965-05-14,S,M,eugene10@adventure-works.com,"$60,000",3,Bachelors,Professional,N
2,11002,MR.,RUBEN,TORRES,1965-08-12,M,M,ruben35@adventure-works.com,"$60,000",3,Bachelors,Professional,Y
3,11003,MS.,CHRISTY,ZHU,1968-02-15,S,F,christy12@adventure-works.com,"$70,000",0,Bachelors,Professional,N
4,11004,MRS.,ELIZABETH,JOHNSON,1968-08-08,S,F,elizabeth5@adventure-works.com,"$80,000",5,Bachelors,Professional,Y


In [260]:
customers['annual_income'] = (
    customers['annual_income']
    .str.replace(r'[$,]',"",regex=True)
)

In [261]:
customers['annual_income'] = pd.to_numeric(
    customers['annual_income'],
    errors='coerce'
    )


In [262]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18148 entries, 0 to 18147
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   customer_key     18148 non-null  int64         
 1   prefix           18018 non-null  object        
 2   first_name       18148 non-null  object        
 3   last_name        18148 non-null  object        
 4   birth_date       18148 non-null  datetime64[ns]
 5   marital_status   18148 non-null  object        
 6   gender           18018 non-null  object        
 7   email_address    18148 non-null  object        
 8   annual_income    18148 non-null  int64         
 9   total_children   18148 non-null  int64         
 10  education_level  18148 non-null  object        
 11  occupation       18148 non-null  object        
 12  home_owner       18148 non-null  object        
dtypes: datetime64[ns](1), int64(3), object(9)
memory usage: 1.8+ MB


In [263]:
customers['prefix'] = (
    customers['prefix']
    .fillna('Unknown')
)

In [264]:
customers['gender'] = (
    customers['gender']
    .fillna('Unknown')
)

In [265]:
customers[["prefix", "gender"]].isnull().sum()

prefix    0
gender    0
dtype: int64

In [266]:
customers.dtypes

customer_key                int64
prefix                     object
first_name                 object
last_name                  object
birth_date         datetime64[ns]
marital_status             object
gender                     object
email_address              object
annual_income               int64
total_children              int64
education_level            object
occupation                 object
home_owner                 object
dtype: object

In [267]:
text_columns = customers.select_dtypes(include='object').columns

for col in text_columns:
    customers[col]=customers[col].str.strip()

In [268]:
customers['first_name'] = customers['first_name'].str.title().astype('string')
customers['last_name'] = customers['last_name'].str.title().astype('string')

In [269]:
customers['marital_status'] = customers['marital_status'].replace({"M":"Married","S":"Single"}).astype("category")

In [270]:
customers['gender']= customers['gender'].replace({'M':'Male','F':'Female'}).astype("category")

# 5. Data Validation

In [271]:
customers['gender'].value_counts(dropna=False)

gender
Male       9126
Female     8892
Unknown     130
Name: count, dtype: int64

In [272]:
customers.head()

,customer_key,prefix,first_name,last_name,birth_date,marital_status,gender,email_address,annual_income,total_children,education_level,occupation,home_owner
0,11000,MR.,Jon,Yang,1966-04-08,Married,Male,jon24@adventure-works.com,90000,2,Bachelors,Professional,Y
1,11001,MR.,Eugene,Huang,1965-05-14,Single,Male,eugene10@adventure-works.com,60000,3,Bachelors,Professional,N
2,11002,MR.,Ruben,Torres,1965-08-12,Married,Male,ruben35@adventure-works.com,60000,3,Bachelors,Professional,Y
3,11003,MS.,Christy,Zhu,1968-02-15,Single,Female,christy12@adventure-works.com,70000,0,Bachelors,Professional,N
4,11004,MRS.,Elizabeth,Johnson,1968-08-08,Single,Female,elizabeth5@adventure-works.com,80000,5,Bachelors,Professional,Y


In [273]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18148 entries, 0 to 18147
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   customer_key     18148 non-null  int64         
 1   prefix           18148 non-null  object        
 2   first_name       18148 non-null  string        
 3   last_name        18148 non-null  string        
 4   birth_date       18148 non-null  datetime64[ns]
 5   marital_status   18148 non-null  category      
 6   gender           18148 non-null  category      
 7   email_address    18148 non-null  object        
 8   annual_income    18148 non-null  int64         
 9   total_children   18148 non-null  int64         
 10  education_level  18148 non-null  object        
 11  occupation       18148 non-null  object        
 12  home_owner       18148 non-null  object        
dtypes: category(2), datetime64[ns](1), int64(3), object(5), string(2)
memory usage: 1.6+ MB


In [274]:
customers.describe()

,customer_key,birth_date,annual_income,total_children
count,18148.000000,18148,18148.000000,18148.000000
mean,20247.470465,1962-01-29 07:41:05.329512864,57269.120564,1.843619
min,11000.000000,1910-08-13 00:00:00,10000.000000,0.000000
25%,15626.750000,1954-09-11 18:00:00,30000.000000,0.000000
50%,20252.500000,1963-08-20 00:00:00,60000.000000,2.000000
75%,24868.250000,1970-09-26 00:00:00,70000.000000,3.000000
max,29483.000000,1980-12-26 00:00:00,170000.000000,5.000000
std,5336.245656,NaN,32236.535573,1.613203


In [275]:
customers.duplicated().sum()

np.int64(0)

In [276]:
customers.isnull().sum()

customer_key       0
prefix             0
first_name         0
last_name          0
birth_date         0
marital_status     0
gender             0
email_address      0
annual_income      0
total_children     0
education_level    0
occupation         0
home_owner         0
dtype: int64

# 6 Feature Engineering

Additional age attributes are created from the `Birth_Date` column to support age-based analysis in SQL and Power BI.

These derived columns improve reporting by enabling analysis by age.

In [277]:
customers['age']=(
    pd.Timestamp.today().year - customers['birth_date'].dt.year
)

# 6. Export Clean Dataset

In [278]:
customers.to_csv(
    "c:/Users/singh/OneDrive/Desktop/AdventureWorks_Analytics/data/cleaned/customers_clean.csv",
    index=False
)